hail/python/hail/ir/ir.py

```diff
@@ -2494,6 +2494,12 @@ def delete(env, name):
         return ApplyAggOp(ir.agg_op,
                           subst_init_op_args,
                           subst_seq_op_args)
+    elif isinstance(ir, AggArrayPerElement):
+        return AggArrayPerElement(_subst(ir.array, delete(agg_env, ir.element_name)),
+                                  ir.element_name,
+                                  ir.index_name,
+                                  ir.agg_ir,
```

Comment:
hmm, I think this is wrong. I think it should be:

```python
        return AggArrayPerElement(_subst(ir.array, env2=agg_env, agg_env2=None),
                                  ir.element_name,
                                  ir.index_name,
                                  _subst(ir.agg_ir, env2=_subst(delete(env, ir.index_name), delete(agg_env, ir.element_name),
                                  ir.is_scan)
```

Reply:
Can you explain to me how env vs agg_env works? Like, why does `ir.array`'s `env2` become an `agg_env`?

Reply:
we have three variable scopes: eval, scan, and agg. Suppose we have a node like TableAggregate, that supports aggregation:
```python
TableAggregate
  table ir
  ApplyBinaryPrimOp
    +
    Ref x
    ApplyAggOp(Sum(), ... seqOpArgs = Seq( Ref y )
```

In the above, the `x` variable is in the eval scope, and `y` is in the agg scope. When we're traversing nodes with data structures that track variables defined/referenced, agg nodes need to do extra work to move variables between scopes. ApplyAggOp takes the agg env coming into it, and uses that agg env as the evaluation env for its seqOp args.

AggArrayPerElement uses the agg environment for its `array` argument, and passes down both an eval and agg env to its `agg_ir` child.

Reply:
does this help?

Reply:
I think I'm following, though I don't know what the `index_name` is for in the code you wrote above.

Reply:
Also, in this line: ```                                  
_subst(ir.agg_ir, env2=_subst(delete(env, ir.index_name), delete(agg_env, ir.element_name)
```
you're calling `_subst` on an `env` instead of an `ir`

Reply:
@tpoterba

Reply:
index_name is a variable exposed in the evaluation scope of agg_ir (the array element index) that can be used in the aggregation result operation.

I think in the above, I added an extra _subst around the env which can be deleted